File: src/Gower.CSharp/Library.cs
namespace SFX
{
    using System;
    using System.Linq;
    using System.Collections.Generic;
    using static System.Double;
    using static System.Math;

    public class GowerConfiguration<T>
    {
        public List<Func<T, bool>> AsymmetricDichotomous { get; set; }
        public List<Func<T, bool>> SymmetricDichotomous { get; set; }
        public List<Func<T, int>> Qualitative { get; set; }
        public List<Func<T, double>> Quantitative { get; set; }
    }

    public static class Gower
    {
        internal static readonly (double, bool) _0_0 = (0.0, false);
        internal static readonly (double, bool) _0_1 = (0.0, true);
        internal static readonly (double, bool) _1_0 = (1.0, false);
        internal static readonly (double, bool) _1_1 = (1.0, true);

        internal static bool IsNumber(double x) =>
            !IsNaN(x) &&
            !IsPositiveInfinity(x) &&
            !IsNegativeInfinity(x) &&
            x != Epsilon &&
            x != MinValue &&
            x != MaxValue;

        internal static (double, bool) DichotomousAsymmetricSimilarity(bool x, bool y) =>
            x && y ? _1_1 : (x != y ? _0_1 : _0_0);
        internal static (double, bool) DichotomousSymmetricSimilarity(bool x, bool y) =>
            x == y ? _1_1 : _0_1;
        internal static (double, bool) QualitativeSimilarity<T>(T x, T y)
            where T : IEquatable<T> =>
            x.Equals(y) ? _1_1 : _0_1;
        internal static (double, bool) QuantitativeSimilarity(double x, double y, double range) 
        {
            if (range == 0.0)
                return (NaN, false);
            if (IsNumber(x) && IsNumber(y) && IsNumber(range))
                return (1.0 - Abs(x - y) / range, true);
            else return (NaN, false);
        }
        internal static (double, bool) GetRange(IEnumerable<(double, bool)> x)
        {
            var valid = x.Where(y => IsNumber(y.Item1) && y.Item2).Select(y => y.Item1);
            if (!valid.Any())
                return (NaN, false);
            var first = valid.First();
            var (a, b) =
                valid
                .Skip(1)
                .Aggregate((Min: first, Max: first), (x, y) => (Min(x.Min, y), Max(x.Max, y)));
            return (b - a, true);
        }

        public static GowerConfiguration<T> 
            WithAsymmetricDichotomousCharacter<T>(this GowerConfiguration<T> conf, 
            Func<T, bool> character)
        {
            if (character is null)
                throw new ArgumentNullException(nameof(character));
            var result = conf ?? new GowerConfiguration<T>();
            if (result.AsymmetricDichotomous is null)
                result.AsymmetricDichotomous = new List<Func<T, bool>> { character };
            else result.AsymmetricDichotomous.Add(character);
            return result;
        }

        public static GowerConfiguration<T>
            WithSymmetricDichotomousCharacter<T>(this GowerConfiguration<T> conf,
            Func<T, bool> character)
        {
            if (character is null)
                throw new ArgumentNullException(nameof(character));
            var result = conf ?? new GowerConfiguration<T>();
            if (result.SymmetricDichotomous is null)
                result.SymmetricDichotomous = new List<Func<T, bool>> { character };
            else result.SymmetricDichotomous.Add(character);
            return result;
        }

        public static GowerConfiguration<T>
            WithQualitativeCharacter<T>(this GowerConfiguration<T> conf,
            Func<T, int> character)
        {
            if (character is null)
                throw new ArgumentNullException(nameof(character));
            var result = conf ?? new GowerConfiguration<T>();
            if (result.Qualitative is null)
                result.Qualitative = new List<Func<T, int>> { character };
            else result.Qualitative.Add(character);
            return result;
        }

        public static GowerConfiguration<T>
            WithQuantitativeCharacter<T>(this GowerConfiguration<T> conf,
            Func<T, double> character)
        {
            if (character is null)
                throw new ArgumentNullException(nameof(character));
            var result = conf ?? new GowerConfiguration<T>();
            if (result.Quantitative is null)
                result.Quantitative = new List<Func<T, double>> { character };
            else result.Quantitative.Add(character);
            return result;
        }

        internal static (double, bool) GetSij(IEnumerable<(double Value, bool IsValid)> x)
        {
            static double AddIf(double x, double y, bool flag) => flag ? x + y : x;
            static int IncIf(int x, bool flag) => flag ? x + 1 : x;
            var (a, b) =
                x
                .Where(y => IsNumber(y.Value) && y.IsValid)
                .Aggregate((Result: 0.0, Count: 0), (w, z) => (AddIf(w.Result, z.Value, z.IsValid), IncIf(w.Count, z.IsValid)));
            if (b == 0)
                return (NaN, false);
            else return (a / b, true);
        }

        internal static Func<T, T, (double, bool)> CreateAsymmetricDichotomous<T>(Func<T, bool> c) =>
            (x, y) => DichotomousAsymmetricSimilarity(c(x), c(y));
        internal static Func<T, T, (double, bool)> CreateSymmetricDichotomous<T>(Func<T, bool> c) =>
            (x, y) => DichotomousSymmetricSimilarity(c(x), c(y));
        internal static Func<T, T, (double, bool)> CreateQualitative<T>(Func<T, int> c) =>
            (x, y) => QualitativeSimilarity(c(x), c(y));
        internal static Func<T, T, double, (double, bool)> CreateQuantitative<T>(Func<T, double> c) =>
            (x, y, range) => QuantitativeSimilarity(c(x), c(y), range);

        public static Func<T, T, IEnumerable<double>, (double Result, bool IsValid)> FromConfig<T>(this GowerConfiguration<T> conf)
        {
            var empty = Enumerable.Empty<Func<T, T, (double, bool)>>();
            var asymmetricDich =
                conf.AsymmetricDichotomous is null ? empty :
                conf.AsymmetricDichotomous
                .Select(x => CreateAsymmetricDichotomous(x));
            var symmetricDich =
                conf.SymmetricDichotomous is null ? empty :
                conf.SymmetricDichotomous
                .Select(x => CreateSymmetricDichotomous(x));
            var qual =
                conf.Qualitative is null ? empty :
                conf.Qualitative
                .Select(x => CreateQualitative(x));
            var quan =
                conf.Quantitative is null ? Enumerable.Empty<Func<T, T, double, (double, bool)>>() :
                conf.Quantitative
                .Select(x => CreateQuantitative(x));
            return (x, y, ranges) =>
            {
                var a = asymmetricDich.Select(w => w(x, y));
                var b = symmetricDich.Select(w => w(x, y));
                var c = qual.Select(w => w(x, y));
                var d = quan.Count() > 0 && ranges.Count() == quan.Count() ?
                    Enumerable.Range(0, quan.Count())
                    .Select(n => (q: quan.ElementAt(n), r: ranges.ElementAt(n)))
                    .Select(f => f.q(x, y, f.r)) :
                    Enumerable.Empty<(double, bool)>();
                return GetSij(a.Concat(b).Concat(c).Concat(d));
            };
        }
    }
}

File: src/Gower/Library.fs
module SFX.Gower

let internal _0_0 = (0.0, false)
let internal _0_1 = (0.0, true)
let internal _1_0 = (1.0, false)
let internal _1_1 = (1.0, true)

let (|NaN|PositiveInfinity|NegativeInfinity|Epsilon|MinValue|MaxValue|Value|) x =
    if System.Double.IsNaN(x) then NaN
    else if System.Double.IsPositiveInfinity(x) then PositiveInfinity
    else if System.Double.IsNegativeInfinity(x) then NegativeInfinity
    else if x = System.Double.Epsilon then Epsilon
    else if x = System.Double.MinValue then MinValue
    else if x = System.Double.MaxValue then MaxValue
    else Value

let internal isNumber x =
    match x with
    | Value -> true
    | _ -> false
let internal dichotomousAsymmetricSimilarity x y =
    match x, y with 
    | true, true -> _1_1
    | true, false | false, true -> _0_1
    | _ -> _0_0
let internal dichotomousSymmetricSimilarity x y =
    match x, y with 
    | true, true | false, false -> _1_1
    | true, false | false, true -> _0_1
let internal qualitativeSimilarity x y = 
    if x = y then _1_1 else _0_1
let internal quantitativeSimilarity x y range =
    if range = 0.0 then (nan, false)
    else 
        match x, y, range with
        | Value, Value, Value -> (1.0 - (abs (x-y) / range), true)
        | _ -> (nan, false)
let internal getRange x =
    let valid =
        x |>
        Seq.filter (fun (value, flag) -> value |> isNumber && flag) |>
        Seq.map fst
    if valid |> Seq.isEmpty then (nan, false)
    else 
        let first = valid |> Seq.head
        let (a, b) = 
            valid |>
            Seq.skip 1 |>
            Seq.fold (fun (x, y) x' -> (min x x', max y x')) (first, first)
        (b-a, true)

type DichotomousCharacter<'a> = 'a -> bool
type QualitativeCharacter<'a> = 'a -> int
type QuantitativeCharacter<'a> = 'a -> float
type GowerConfiguration<'a> = {
    AsymmetricDichotomous: DichotomousCharacter<'a> list option
    SymmetricDichotomous: DichotomousCharacter<'a> list option
    Qualitative: QualitativeCharacter<'a> list option
    Quantitative: QuantitativeCharacter<'a> list option
}
let createGowerConfiguration() = {
    AsymmetricDichotomous = None
    SymmetricDichotomous = None
    Qualitative = None
    Quantitative = None
}
let withAsymmetricDichotomousCharacter c conf =
    match conf.AsymmetricDichotomous with
    | Some l -> {conf with AsymmetricDichotomous = c::l |> Some}
    | None -> {conf with AsymmetricDichotomous = [c] |> Some}
let withSymmetricDichotomousCharacter c conf =
    match conf.SymmetricDichotomous with
    | Some l -> {conf with SymmetricDichotomous = c::l |> Some}
    | None -> {conf with SymmetricDichotomous = [c] |> Some}
let withQualitativeCharacter c conf =
    match conf.Qualitative with
    | Some l -> {conf with Qualitative = c::l |> Some}
    | None -> {conf with Qualitative = [c] |> Some}
let withQuantitativeCharacter c conf =
    match conf.Quantitative with
    | Some l -> {conf with Quantitative = c::l |> Some}
    | None -> {conf with Quantitative = [c] |> Some}

let internal getSij x =
    let add_if x y flag = if flag then x+y else x
    let inc_if x flag = if flag then x+1 else x
    let (a, b) = 
        x |> 
        List.filter (fun (value, flag) -> value |> isNumber && flag) |>
        List.fold (fun (a', b') (a'',b'') -> (add_if a' a'' b'', inc_if b' b'')) (0.0, 0)
    if b = 0 then (nan, false)
    else (a/(float b), true)
    
let createAsymmetricDichotomous (c: DichotomousCharacter<'a>) =
    fun x y -> dichotomousAsymmetricSimilarity (x |> c) (y |> c)
let createSymmetricDichotomous (c: DichotomousCharacter<'a>) =
    fun x y -> dichotomousSymmetricSimilarity (x |> c) (y |> c)
let createQualitative (c: QualitativeCharacter<'a>) =
    fun x y -> qualitativeSimilarity (x |> c) (y |> c)
let createQuantitative (c: QuantitativeCharacter<'a>) =
    fun x y range -> quantitativeSimilarity (x |> c) (y |> c) range
let fromConfig conf =
    let asymmetricDich =
        match conf.AsymmetricDichotomous with
        | Some l -> l |> List.map createAsymmetricDichotomous
        | _ -> []
    let symmetricDich =
        match conf.SymmetricDichotomous with
        | Some l -> l |> List.map createSymmetricDichotomous
        | _ -> []
    let qual =
        match conf.Qualitative with
        | Some l -> l |> List.map createQualitative
        | _ -> []
    let quan =
        match conf.Quantitative with
        | Some l -> l |> List.map createQuantitative
        | _ -> []
    fun x y ranges ->
        let a = asymmetricDich |> List.map (fun c -> c x y)
        let b = symmetricDich |> List.map (fun c -> c x y)
        let c = qual |> List.map (fun c -> c x y)
        let d =
            match quan |> List.length, ranges |> List.length with
            | q, r when q > 0 && q = r -> (quan, ranges) ||> List.map2 (fun c range -> c x y range)
            | _ -> []
        a @ b @ c @ d |> getSij
